src/audio/pipeline.c

```diff
@@ -616,6 +616,30 @@ static void pipeline_comp_trigger_sched_comp(struct pipeline *p,
 	list_item_append(&p->list, &ctx->pipelines);
 }
 
+/* if capture START/RELEASE trigger originated on dailess pipeline and reached
+ * inactive pipeline as it's source, then we indicate that it's blocked.
+ */
+static inline int
+pipeline_depender_is_blocked(struct comp_dev *dependee,
+			     struct pipeline_walk_context *ctx,
+			     int dir)
+{
+	struct pipeline_data *ppl_data = ctx->comp_data;
+	int is_trigger_start_or_release = ppl_data->cmd == COMP_TRIGGER_START ||
+					  ppl_data->cmd == COMP_TRIGGER_RELEASE;
+	int is_capture = dir == SOF_IPC_STREAM_CAPTURE;
+	struct comp_dev *pipe_source = ppl_data->start->pipeline->source_comp;
+	int has_dai_source = pipe_source &&
+			     dev_comp_type(pipe_source) == SOF_COMP_DAI;
+	int is_dependee_active = dependee->state == COMP_STATE_ACTIVE;
+
+	if (is_capture && is_trigger_start_or_release && !has_dai_source &&
+	    !is_dependee_active)
+		return 1;
+
```

Comment:
Far easier to read if done like this
```c
/* only applies to capture pipelines */
if (!is_capture)
    return 0;
/* not valid if pipeline not running */
if (!is_pipeline_running)
    return 0;
/// and so on.....
```

Reply:
shoule we the use intermediate variables, or leave the explanation to the comments and plug in the logic that is in the assignments right now?